File: fifteen-exercises/spiral-matrix.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <functional>
#include <iterator>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>> &matrix) {
        if (matrix.empty()) return vector<int>();
        const int x_max = matrix[0].size();
        const int y_max = matrix.size();
        int x = -1, y = 0;
        const int dx[] = {1, 0, -1, 0};
        const int dy[] = {0, 1, 0, -1};
        int dir = 0;
        vector<int> res;
        while (true) {
            while (true) {
                x = x + dx[dir];
                y = y + dy[dir];
                if (x < 0 or x >= x_max or y < 0 or y >= y_max or matrix[y][x] == INT_MAX) {
                    x -= dx[dir];
                    y -= dy[dir];
                    break;
                }
                res.emplace_back(matrix[y][x]);
                matrix[y][x] = INT_MAX;
            }
            if (res.size() == x_max * y_max) {
                break;
            }
            dir = (dir + 1) % 4;
        }
        return res;
    }
};

int main() {
    Solution sln = Solution();
    vector<vector<int>> test = {{1, 2, 3},
                                {4, 5, 6},
                                {7, 8, 9}};
    vector<int> xs = sln.spiralOrder(test);
    copy(xs.begin(), xs.end(), ostream_iterator<int>(cout, ", "));
    cout << endl;
    return 0;
}

File: fifteen-exercises/spiral-matrix.rs
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        if matrix.len() == 0 { return Vec::new() }
        let (x_max, y_max) = (matrix[0].len(), matrix.len());
        let mut vis = vec![vec![false; x_max]; y_max]; // 也可以直接用可变matrix记录
        let (mut x, mut y) = (-1, 0);
        let dx = vec![1i32, 0, -1, 0];
        let dy = vec![0i32, 1, 0, -1];
        let mut dir = 0usize;
        let mut res = Vec::new();
        loop {
            loop {
                x = x + dx[dir];
                y = y + dy[dir];
                if x < 0 || x >= x_max as i32 || y < 0 || y >= y_max as i32 || vis[y as usize][x as usize] {
                    x -= dx[dir];
                    y -= dy[dir]; // 退回去
                    break;
                }
                res.push(matrix[y as usize][x as usize]);
                vis[y as usize][x as usize] = true;
            }
            if res.len() == x_max * y_max {
                break;
            }
            dir = (dir + 1) % 4;
        }
        res
    }
}

struct Solution;

fn main() {
    println!("{:?}", Solution::spiral_order(vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]]));
    println!("ok")
}
